lib/websocket/connection.js

```diff
@@ -0,0 +1,409 @@
+
+      if (secProtocol !== null && secProtocol !== request.headersList.get('Sec-WebSocket-Protocol')) {
+        failWebsocketConnection(controller, response.socket)
+        return
+      }
+
+      // processResponse is called when the "response’s header list has been received and initialized."
+      // once this happens, the connection is open
+      ws[kResponse] = response
+
+      whenConnectionEstablished(ws)
+
+      receiveData(ws)
+
+      socketClosed(ws)
+    }
+  })
+
+  return controller
+}
+
+/**
+ * @param {import('../fetch/index').Fetch} controller
+ * @param {import('net').Socket|import('tls').TLSSocket|undefined} socket
+ */
+function failWebsocketConnection (controller, socket) {
+  controller.abort()
+
+  if (socket && !socket.destroyed) {
+    socket.destroy()
+  }
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function whenConnectionEstablished (ws) {
+  const { [kResponse]: response } = ws
+
+  // 1. Change the ready state to OPEN (1).
+  ws[kReadyState] = states.OPEN
+
+  // 2. Change the extensions attribute’s value to the extensions in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-9.1
+  const extensions = response.headersList.get('sec-websocket-extensions')
+
+  if (extensions !== null) {
+    ws[kExtensions] = extensions
+  }
+
+  // 3. Change the protocol attribute’s value to the subprotocol in use, if
+  //    it is not the null value.
+  // https://datatracker.ietf.org/doc/html/rfc6455#section-1.9
+  const protocol = response.headersList.get('sec-websocket-protocol')
+
+  if (protocol !== null) {
+    ws[kProtocol] = protocol
+  }
+
+  // 4. Fire an event named open at the WebSocket object.
+  fireEvent('open', ws)
+}
+
+/**
+ * @see https://websockets.spec.whatwg.org/#feedback-from-the-protocol
+ * @param {import('./websocket').WebSocket} ws
+ */
+function receiveData (ws) {
+  const { [kResponse]: response } = ws
+
+  /** @type {WebsocketFrame|undefined} */
+  let frame
+
+  response.socket.on('data', (chunk) => {
+    const receivedFrame = WebsocketFrame.from(chunk)
+
+    if (!frame) {
+      frame = receivedFrame
+      frame.terminated = frame.fin // message complete
+    } else {
+      // A fragmented message consists of a single frame with the FIN bit
+      // clear and an opcode other than 0, followed by zero or more frames
+      // with the FIN bit clear and the opcode set to 0, and terminated by
+      // a single frame with the FIN bit set and an opcode of 0.
+
+      if (receivedFrame.fin && receivedFrame.opcode === opcodes.CONTINUATION) {
+        frame.terminated = true
+      }
+
+      frame.addFrame(receivedFrame.data)
+    }
+
+    // If a control frame (Section 5.5) is
+    // received, the frame MUST be handled as defined by Section 5.5.
+    if (frame.opcode === opcodes.PING) {
+      handlePing(response.socket, ws)
```

Comment:
This should be done after all frame data is read because the pong frame must have the same data of the ping frame being replied to.

Reply:
The implementation here is currently lacking - the server doesn't necessarily need to send the same data it's responding to, just the data from the latest ping it received. The server *can* respond to both pings, but it's not required.

```
   If an endpoint receives a Ping frame and has not yet sent Pong
   frame(s) in response to previous Ping frame(s), the endpoint MAY
   elect to send a Pong frame for only the most recently processed Ping
   frame.
```

Also, a server doesn't need to send one right away, just when it's "practical" to do so. I was thinking of adding a ping queue with a ms timestamp and deleting all earlier pings, but as you said, an evil/bad server just might not send any pongs.

Reply:
I was referring to this https://www.rfc-editor.org/rfc/rfc6455#section-5.5.3

> A Pong frame sent in response to a Ping frame must have identical
   "Application data" as found in the message body of the Ping frame
   being replied to.

You can't send a pong without reading the ping data.

Reply:
Currently we send an empty ping frame so the payload length of the pong frame must be 0 (validation is done to check). Not the greatest solution ever, but it should be okay?

Reply:
Interesting, I was confusing "endpoint" in the rfc with "server"

Reply:
Just wondering, how does ws handle pongs? What if the user pings twice and only receives a pong for the last ping? What if a server sends pongs for both? I have an idea of how to fix it, but it seems as though ws 

> the endpoint MAY
   elect to send a Pong frame for only the most recently processed Ping
   frame.

```js
const ws = new WebSocket('...')

ws.ping('...')
ws.ping('...')

// assume that the server either responds to both pings, or only 1 of them

```

Reply:
> What if the user pings twice and only receives a pong for the last ping? 

That's ok, it does not matter, pongs need no special handling.

> What if a server sends pongs for both?

Same answer.

> I have an idea of how to fix it, but it seems as though ws handles this much easier than I'm thinking. Am I overthinking this? Am I overthinking this?

I think so, `ws` responds to all pings, not just the most recent one. If the implementation sends a pong for every received ping there is no `If an endpoint receives a Ping frame and has not yet sent Pong frame(s) in response to previous Ping frame(s)` case.

Reply:
> pongs need no special handling

I guess that's where my issue was. I was assuming that I'd need to validate the application data of the pong frame, to ensure it was replying to a ping we sent. Although it gets complicated since pong frames could also be unsolicited.